tests/conftest.py

```diff
@@ -68,3 +82,86 @@ async def stream_writer(
     async_buffers: tuple[object, asyncio.StreamWriter],
 ) -> asyncio.StreamWriter:
     return async_buffers[1]
+
+
+class JavaTester:
+    class _Encoder(JSONEncoder):
+        def default(self, o: Any) -> Any:
+            if dataclasses.is_dataclass(o):
+                return self._replace_tzaware_nulls(dataclasses.asdict(o))
+            if isinstance(o, timedelta):
+                return round(o.total_seconds() * 1000)
+            if isinstance(o, datetime):
+                return round(o.timestamp() * 1000)
+            if isinstance(o, UUID):
+                return str(o)
+            if isinstance(o, bytes):
+                return base64.b64encode(o).decode("utf-8")
+            return super().default(o)
+
+        def _replace_tzaware_nulls(self, o: Any) -> Any:
+            if isinstance(o, dict):
+                result = {}
+                for k, v in o.items():
+                    if k == "log_append_time" and v is None:
+                        result[k] = -1
+                    else:
+                        result[k] = self._replace_tzaware_nulls(v)
+                return result
+            elif isinstance(o, list):
+                return [self._replace_tzaware_nulls(e) for e in o]
+            elif isinstance(o, tuple):
+                return tuple(self._replace_tzaware_nulls(e) for e in o)
+            else:
+                return o
+
+    def __init__(self) -> None:
+        cmd = [
+            "docker",
+            "run",
+            "--rm",
+            "-i",
+            "java_tester",
+            "/java_tester/bin/java_tester",
+        ]
```

Comment:
question(non-blocking): Is it possible we could run this as a service managed in the docker-compose file? Small thing that might not be worth the effort, but it would allow a slightly simpler dev workflow as it can automatically build the image.

Reply:
Yes, good idea. Implemented